lib/time/time.go

```diff
@@ -0,0 +1,548 @@
+func (d *Duration) Unpack(v starlark.Value) error {
+	switch x := v.(type) {
+	case Duration:
+		*d = x
+		return nil
+	case starlark.Int:
+		i, ok := x.Int64()
+		if !ok {
+			return fmt.Errorf("int value out of range (want signed 64-bit value)")
+		}
+		*d = Duration(i)
+		return nil
+	case starlark.String:
+		dur, err := time.ParseDuration(string(x))
+		if err != nil {
+			return err
+		}
+
+		*d = Duration(dur)
+		return nil
+	}
+
+	return fmt.Errorf("cannot convert %s to %s", v.Type(), d.Type())
+}
+
+// String implements the Stringer interface.
+func (d Duration) String() string { return time.Duration(d).String() }
+
+// Type returns a short string describing the value's type.
+func (d Duration) Type() string { return "time.duration" }
+
+// Freeze renders Duration immutable. required by starlark.Value interface
+// because duration is already immutable this is a no-op.
+func (d Duration) Freeze() {}
+
+// Hash returns a function of x such that Equals(x, y) => Hash(x) == Hash(y)
+// required by starlark.Value interface.
+func (d Duration) Hash() (uint32, error) {
+	return starlark.MakeInt64(int64(d)).Hash()
+}
+
+// Truth returns the truth value of an object required by starlark.Value
+// interface.
+func (d Duration) Truth() starlark.Bool { return d != 0 }
+
+// Attr gets a value for a string attribute, implementing dot expression support
+// in starklark. required by starlark.HasAttrs interface.
+func (d Duration) Attr(name string) (starlark.Value, error) {
+	switch name {
+	case "hours":
+		return starlark.Float(time.Duration(d).Hours()), nil
+	case "minutes":
+		return starlark.Float(time.Duration(d).Minutes()), nil
+	case "seconds":
+		return starlark.Float(time.Duration(d).Seconds()), nil
+	case "nanoseconds":
+		return starlark.MakeInt64(time.Duration(d).Nanoseconds()), nil
+	}
+	return nil, fmt.Errorf("unrecognized %s attribute %q", d.Type(), name)
+}
+
+// AttrNames lists available dot expression strings. required by
+// starlark.HasAttrs interface.
+func (d Duration) AttrNames() []string {
+	return []string{
+		"hours",
+		"minutes",
+		"seconds",
+		"nanoseconds",
+	}
+}
+
+// Binary implements binary operators, which satisfies the starlark.HasBinary
+// interface. operators:
+//    duration + duration = duration
+//    duration + int = duration
+//    duration + time = time
+//    duration - duration = duration
+//    duration - int = duration
+//    duration - time = duration
+//    duration == duration = boolean
+//    duration < duration = boolean
+func (d Duration) Binary(op syntax.Token, yV starlark.Value, _ starlark.Side) (starlark.Value, error) {
+	x := time.Duration(d)
+
+	switch op {
+	case syntax.PLUS:
+		switch y := yV.(type) {
+		case Duration:
+			return Duration(x + time.Duration(y)), nil
+		case starlark.Int:
+			i, ok := y.Int64()
+			if !ok {
+				return nil, fmt.Errorf("int value out of range (want signed 64-bit value)")
+			}
+			return Duration(x + time.Duration(i)), nil
```

Comment:
I don't think we should allow this operation. Duration * int makes sense, but Duration + int does not. (It does in Go, because Go's time.Duration exposes its representation as an int64 of nanoseconds.) Users can always write `duration + 1 * time.nanosecond`, which is clear and unambiguous.